docs/api/Inside-the-Template-Engine.md

```diff
@@ -0,0 +1,340 @@
+# Inside the Template Engine
+
+The template engine is composed of several different subsystems, which are
+designed to separate gathering templates, instantiating template and processing
+templates.
+
+# Overview of the subsystems
+
+## [IEngineEnvironmentSettings](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IEngineEnvironmentSettings.cs)
+
+This is responsible for holding all properties of environment.
+
+Template engine provides default implementation:
+[EngineEnvironmentSettings](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Edge/EngineEnvironmentSettings.cs).
+
+It has single constructor that accepts ITemplateEngineHost and set up settings
+based on TemplateEngine.Edge implementation of component manager, paths, and
+environment. All this can also be passed in as optional parameters.
+
+### [ITemplateEngineHost](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs)
+
+The interface is responsible for providing host information to template engine,
+managing file system and do the logging. Host may also provide default parameter
+values for template via TryGetHostParamDefault method.
+
+The applications using template engine are often called as “hosts”.
+dotnet/templating repo is managing one of such hosts: dotnet new CLI, which is
+part of dotnet SDK.
+
+Main host properties:
+
+-   Identifier – unique name of the host
+
+-   Version – version of the host
+
+Those properties are used to identify host to various built-in components
+explained below.
+
+If you are considering using the template engine core, you need to make
+implementation of this interface representing your host. Template engine
+provides default implementation:
+[DefaultTemplateEngineHost](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Edge/DefaultTemplateEngineHost.cs).
+
+#### IPhysicalFileSystem
```

Comment:
Missing link.
Is the header depth correct?

Reply:
Added link, depth is correct